### ISSUE TITLE
Typings for Typescript projects?

### ISSUE THREAD

**Issue by creaux** [bootstrap-styled/bootstrap-styled#83]

**Comment by kopax** [comment]
PR welcome

**Comment by kopax** [comment]
@creaux would you have time and be interested in adding support of typescript in V4 + Bootstrap Styled?

**Comment by creaux** [comment]
@kopax I can’t help you on this at this moment. So sorry.

**Comment by tim-stasse** [comment]
Hi @kopax,

I've created some PR's adding typescript definition files to the core required packages:

[#136](https://github.com/bootstrap-styled/v4/pull/136)
[#41](https://github.com/bootstrap-styled/provider/pull/41)
[#105](https://github.com/bootstrap-styled/bootstrap-styled/pull/105)
[#38](https://github.com/bootstrap-styled/css-utils/pull/38)
[#30](https://github.com/bootstrap-styled/utils/pull/30)

More work needs to be done before the definitions can be used though. The build process needs to be updated for each project to include the definition file in each projects respective outputs. I'll leave that up to you or anyone else who wants to either add to the PR or as an additional one.

Note that with future changes to the interfaces for any exported variables, functions, components, etc. the respective types will also need to be separately updated as well.

**Comment by kopax** [comment]
> Note that with future changes to the interfaces for any exported variables, functions, components, etc. the respective types will also need to be separately updated as well.

I will add a note early september for this. I have merged  on v4 and bootstrap-styled. I wait for travis being able to merge master or i will do it in september for the others. Is it possible you test v4 with ts with the latest release version?

Thanks a lot for your contribution!

**Comment by kopax** [comment]
All merged, I close this unless feedback.

**Comment by tim-stasse** [comment]
Hi @kopax, yes it's possible to automatically generate these, if you re-write the entire project in typescript. Which is do-able, but also more work up-front (but less work in the long-run in terms of maintaining the types).

I just tested the latest release, which has the type definitions in the source code, however it's not being included in the dist folder of the package artifact, and is therefore not currently working.

I mentioned in the PR's that they should probably need to be explicitly included in the release artifacts. I haven't used rollup before, so I don't understand the exact setup you have, but I expect rollup acts much like webpack, in regards to excluding unused files when traversing the AST tree. Which would explain the declaration file not being included in the dist folder.

**Comment by kopax** [comment]
> Which is do-able, but also more work up-front (but less work in the long-run in terms of maintaining the types).

We should go that way, I'll accept the currency implémentation for ts users but we should migrate our project to typescript. I do not have experience with it so I wait to have some help and testeurs to involve myself in the understanding.

> I just tested the latest release, which has the type definitions in the source code, however it's not being included in the dist folder of the package artifact, and is therefore not currently working.

Thanks for testing,  I'll be able to configure it properly in september. I will need more testeurs or better, a sample project so I can reproduce what you have.

**Comment by aequasi** [comment]
These don't seem to have been published to npm. Probably needs to either be placed in the root dir, instead of `src`, or moved to `lib` after a build

**Comment by aequasi** [comment]
Either way, an update to `package.json` needs to be made to add `types`: https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#including-declarations-in-your-npm-package

**Comment by kopax** [comment]
Babel parameter `--copy-files` should be used in `npm run build:lib` command so it should be in `lib`.

See `package.json` of all projects and https://stackoverflow.com/questions/32642685/babel-cli-copy-nonjs-files


I am not able to edit the project atm, but I can pass PR if you take care of it. Please check all the projects from the list here and do the PR on merge branch.

For dist (rollup) I will have to try in september, and I still need a sample TS project for testing. As I am not experienced with it, it should be done by one of you. 

Thanks

**Comment by kopax** [comment]
I am republishing now with the `--copy-files` on for `lib` distribution. 
I have checked an article https://hackernoon.com/building-and-publishing-a-module-with-typescript-and-rollup-js-faa778c85396 and I wonder if it's possible to just inject the typescript definition without switching the whole project to typescript. Is anybody experience or have proper resource to help me go with that task?

**Comment by aequasi** [comment]
Yes, you can just include the typescript definition, but it does make it harder in the long run, as you have to manually update the types anytime you change anything in the library (as opposed to them being autogenerated)

**Comment by kopax** [comment]
> you can just include the typescript definition

Do you know how to setup rollup to do that?

**Comment by kopax** [comment]
@aequasi, can you please have a look at @vinterdo issue https://github.com/bootstrap-styled/v4/issues/157

> Typings file (index.d.ts) contain invalid typings. Not only they are not compiling (typos, using "bool" instead of "boolean", "function" instead of "Function", but also they prevent me from using bootstrap styled-components (eg. components are not accepting props like a name on button, causing compilation error)
> 
> The temporary workaround is to manually delete typings file from node_modules. It works just fine then. Problem is that this workaround is fine when I'm developing on my pc, pretty unusable on the CI server.

**Comment by vinterdo** [comment]
I would recommend reverting them for now, and later - either consider TypeScript-first approach or maintain a separate, optional repository and package for typings - it might be out of date, but won't break new versions of bootstrap-styled. The current approach is no-go, I can't compile my project with newer versions of bootstrap-styled.

**Comment by kopax** [comment]
@vinterdo the code source and variables has not changed for a while, and the one time I updated variables, I also updated the typing. 

Perhaps I can pass the syntax fix so you can enjoy the typings until we find a proper solution. 
Reverting the whole typings is about the same effort than updating those variables.

If you will can you describe what syntax is wrong, such as:

- `bool` => `boolean`.

In any case, I will wait for more feedback until I found a proper solution.

Meanwhile, if you can't deal with the wrong typings in `node_modules`, I suggest you fix the version before the typescript addition (see changelog for that).

**Comment by ManAnRuck** [comment]
Do you have any updates for this?

**Comment by kopax** [comment]
HI @ManAnRuck , I am just getting started with TypeScript, if one want to migrate with Typings, PR are welcome.

**Comment by krobing** [comment]
Hi there, I want to say that at the moment I have a project that built-in the types definition for the `bootstrap-styled` library and it works for me. This is how the typings file looks: [bootstrap-styled.d.ts](https://gist.github.com/krobing/1abd570633114d20307cf02826296213)

also I've seen another way how other projects that were not built with Typescript have their types definition into the repo: [https://github.com/DefinitelyTyped/DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) and published as npm package, so it would be named `@types/bootstrap-styled`.

This is helpful if it's wishing the ui library (bootstrap-styled) doesn't leans on types definition.
